### ISSUE TITLE
Harbor is not connecting to OIDC (500 error message)

### ISSUE THREAD

**Issue by cmpmohan** [goharbor/harbor#9088]
What can we help you?

We have configured our harbor registry with OIDC endpoint and credentials. harbor is not connecting to OIDC endpoint. when we are clicking Login via OIDC provider option in harbor URL its showing error message “{"code":500,"message":"Internal Server Error"}”.  please help us to fix this issue.

============================================

Sep 13 23:26:31 172.21.0.1 core[16925]: 2019-09-13T17:56:31Z [DEBUG] [/core/filter/security.go:576]: creating local database security context...
Sep 13 23:26:32 172.21.0.1 core[16925]: 2019-09-13T17:56:32Z [ERROR] [/common/utils/oidc/helper.go:148]: Failed to get OAuth configuration, error: failed to create OIDC provider, error: 404 Not Found: <!DOCTYPE html>#015
Sep 13 23:26:32 172.21.0.1 core[16925]: #015
Sep 13 23:26:32 172.21.0.1 core[16925]: #015
Sep 13 23:26:32 172.21.0.1 core[16925]: #015
Sep 13 23:26:32 172.21.0.1 core[16925]: #015
Sep 13 23:26:32 172.21.0.1 core[16925]: #015
Sep 13 23:26:32 172.21.0.1 core[16925]: <html lang="en" dir="ltr">#015
Sep 13 23:26:32 172.21.0.1 core[16925]: <head>#015
Sep 13 23:26:32 172.21.0.1 core[16925]:     <title>Error</title>#015
Sep 13 23:26:32 172.21.0.1 core[16925]:     <base href="https://sso-int.daimler.com/" />#015
Sep 13 23:26:32 172.21.0.1 core[16925]:     <meta name="robots" content="noindex, nofollow" />#015
Sep 13 23:26:32 172.21.0.1 core[16925]:     <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />#015
Sep 13 23:26:32 172.21.0.1 core[16925]:     <meta name="viewport" content="initial-scale=1.0, minimum-scale=1.0, maximum-scale=1.0, user-scalable=no" />#015
Sep 13 23:26:32 172.21.0.1 core[16925]:     <meta http-equiv="x-ua-compatible" content="IE=edge" />#015
Sep 13 23:26:32 172.21.0.1 core[16925]:     <link rel="stylesheet" type="text/css" href="assets/css/main.css" />#015
Sep 13 23:26:32 172.21.0.1 core[16925]: </head>#015
Sep 13 23:26:32 172.21.0.1 core[16925]: #015
Sep 13 23:26:32 172.21.0.1 core[16925]: <body>#015
Sep 13 23:26:32 172.21.0.1 core[16925]:     #015
Sep 13 23:26:32 172.21.0.1 core[16925]: <div class="ping-container">#015
Sep 13 23:26:32 172.21.0.1 core[16925]:     #015
Sep 13 23:26:32 172.21.0.1 core[16925]:     #015
Sep 13 23:26:32 172.21.0.1 core[16925]:     <div class="ping-header">#015
Sep 13 23:26:32 172.21.0.1 core[16925]:         <span class="company-logo"></span>#015
Sep 13 23:26:32 172.21.0.1 core[16925]:         Error#015
Sep 13 23:26:32 172.21.0.1 core[16925]:     </div>#015
Sep 13 23:26:32 172.21.0.1 core[16925]:     #015
Sep 13 23:26:32 172.21.0.1 core[16925]:     <div class="ping-body-container">#015
Sep 13 23:26:32 172.21.0.1 core[16925]:         <div class="section-title">#015
Sep 13 23:26:32 172.21.0.1 core[16925]:             Oops#015
Sep 13 23:26:32 172.21.0.1 core[16925]:         </div>#015
Sep 13 23:26:32 172.21.0.1 core[16925]:         <div class="ping-messages ping-nopad">#015
Sep 13 23:26:32 172.21.0.1 core[16925]:             <div>#015
Sep 13 23:26:32 172.21.0.1 core[16925]:                 Looks like something is not right. Please contact your administrator.#015
Sep 13 23:26:32 172.21.0.1 core[16925]:             </div>#015
Sep 13 23:26:32 172.21.0.1 core[16925]:             <div class="ping-note-text">#015
Sep 13 23:26:32 172.21.0.1 core[16925]:                 404 - Not Found#015
Sep 13 23:26:32 172.21.0.1 core[16925]:             </div>#015
Sep 13 23:26:32 172.21.0.1 core[16925]:         </div>#015
Sep 13 23:26:32 172.21.0.1 core[16925]:     </div> #015
Sep 13 23:26:32 172.21.0.1 core[16925]:     #015
Sep 13 23:26:32 172.21.0.1 core[16925]:     <div class="ping-footer-container">#015
Sep 13 23:26:32 172.21.0.1 core[16925]:         <div class="ping-footer">#015
Sep 13 23:26:32 172.21.0.1 core[16925]:             <div class="ping-credits"></div>#015
Sep 13 23:26:32 172.21.0.1 core[16925]:             <div class="ping-copyright">Copyright © 2003-2019. Ping Identity Corporation. All rights reserved.</div>#015
Sep 13 23:26:32 172.21.0.1 core[16925]:         </div> #015
Sep 13 23:26:32 172.21.0.1 core[16925]:     </div> #015
Sep 13 23:26:32 172.21.0.1 core[16925]:     #015
Sep 13 23:26:32 172.21.0.1 core[16925]: </div> #015
Sep 13 23:26:32 172.21.0.1 core[16925]: #015
Sep 13 23:26:32 172.21.0.1 core[16925]: </body>#015
Sep 13 23:26:32 172.21.0.1 core[16925]: </html>#015
Sep 13 23:26:32 172.21.0.1 core[16925]: 2019-09-13T17:56:32Z [ERROR] [/common/api/base.go:229]: failed to create OIDC provider, error: 404 Not Found: <!DOCTYPE html>#015
Sep 13 23:26:32 172.21.0.1 core[16925]: #015
Sep 13 23:26:32 172.21.0.1 core[16925]: #015
Sep 13 23:26:32 172.21.0.1 core[16925]: #015
Sep 13 23:26:32 172.21.0.1 core[16925]: #015
Sep 13 23:26:32 172.21.0.1 core[16925]: #015
Sep 13 23:26:32 172.21.0.1 core[16925]: <html lang="en" dir="ltr">#015
Sep 13 23:26:32 172.21.0.1 core[16925]: <head>#015
Sep 13 23:26:32 172.21.0.1 core[16925]:     <title>Error</title>#015
Sep 13 23:26:32 172.21.0.1 core[16925]:     <base href="https://sso-int.daimler.com/" />#015
Sep 13 23:26:32 172.21.0.1 core[16925]:     <meta name="robots" content="noindex, nofollow" />#015
Sep 13 23:26:32 172.21.0.1 core[16925]:     <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />#015
Sep 13 23:26:32 172.21.0.1 core[16925]:     <meta name="viewport" content="initial-scale=1.0, minimum-scale=1.0, maximum-scale=1.0, user-scalable=no" />#015
Sep 13 23:26:32 172.21.0.1 core[16925]:     <meta http-equiv="x-ua-compatible" content="IE=edge" />#015
Sep 13 23:26:32 172.21.0.1 core[16925]:     <link rel="stylesheet" type="text/css" href="assets/css/main.css" />#015
Sep 13 23:26:32 172.21.0.1 core[16925]: </head>#015
Sep 13 23:26:32 172.21.0.1 core[16925]: #015
Sep 13 23:26:32 172.21.0.1 core[16925]: <body>#015
Sep 13 23:26:32 172.21.0.1 core[16925]:     #015
Sep 13 23:26:32 172.21.0.1 core[16925]: <div class="ping-container">#015
Sep 13 23:26:32 172.21.0.1 core[16925]:     #015
Sep 13 23:26:32 172.21.0.1 core[16925]:     #015
Sep 13 23:26:32 172.21.0.1 core[16925]:     <div class="ping-header">#015
Sep 13 23:26:32 172.21.0.1 core[16925]:         <span class="company-logo"></span>#015
Sep 13 23:26:32 172.21.0.1 core[16925]:         Error#015
Sep 13 23:26:32 172.21.0.1 core[16925]:     </div>#015
Sep 13 23:26:32 172.21.0.1 core[16925]:     #015
Sep 13 23:26:32 172.21.0.1 core[16925]:     <div class="ping-body-container">#015
Sep 13 23:26:32 172.21.0.1 core[16925]:         <div class="section-title">#015
Sep 13 23:26:32 172.21.0.1 core[16925]:             Oops#015
Sep 13 23:26:32 172.21.0.1 core[16925]:         </div>#015
Sep 13 23:26:32 172.21.0.1 core[16925]:         <div class="ping-messages ping-nopad">#015
Sep 13 23:26:32 172.21.0.1 core[16925]:             <div>#015
Sep 13 23:26:32 172.21.0.1 core[16925]:                 Looks like something is not right. Please contact your administrator.#015
Sep 13 23:26:32 172.21.0.1 core[16925]:             </div>#015
Sep 13 23:26:32 172.21.0.1 core[16925]:             <div class="ping-note-text">#015
Sep 13 23:26:32 172.21.0.1 core[16925]:                 404 - Not Found#015
Sep 13 23:26:32 172.21.0.1 core[16925]:             </div>#015
Sep 13 23:26:32 172.21.0.1 core[16925]:         </div>#015
Sep 13 23:26:32 172.21.0.1 core[16925]:     </div> #015
Sep 13 23:26:32 172.21.0.1 core[16925]:     #015
Sep 13 23:26:32 172.21.0.1 core[16925]:     <div class="ping-footer-container">#015
Sep 13 23:26:32 172.21.0.1 core[16925]:         <div class="ping-footer">#015
Sep 13 23:26:32 172.21.0.1 core[16925]:             <div class="ping-credits"></div>#015
Sep 13 23:26:32 172.21.0.1 core[16925]:             <div class="ping-copyright">Copyright © 2003-2019. Ping Identity Corporation. All rights reserved.</div>#015
Sep 13 23:26:32 172.21.0.1 core[16925]:         </div> #015
Sep 13 23:26:32 172.21.0.1 core[16925]:     </div> #015
Sep 13 23:26:32 172.21.0.1 core[16925]:     #015
Sep 13 23:26:32 172.21.0.1 core[16925]: </div> #015
Sep 13 23:26:32 172.21.0.1 core[16925]: #015
Sep 13 23:26:32 172.21.0.1 core[16925]: </body>#015
Sep 13 23:26:32 172.21.0.1 core[16925]: </html>#015
Sep 13 23:26:32 172.21.0.1 core[16925]: 2019-09-13T17:56:32Z [ERROR] [/common/api/base.go:69]: GET /c/oidc/login failed with error: {"code":500,"message":"Internal Server Error"}
Sep 13 23:26:32 172.21.0.1 core[16925]: 2019/09/13 17:56:32 #033[1;44m[D] [server.go:2619] | 53.244.253.207|#033[41m 500 #033[0m| 1.275203736s|   match|#033[44m GET     #033[0m /c/oidc/login   r:/c/oidc/login#033[0m

**Comment by reasonerjt** [comment]
This looks like a mis-configuration, how did you set your OIDC endpoint URL?

**Comment by cmpmohan** [comment]
We have set OIDC details on our harbor URL, 

Logged on harbor URL ==> go to configuration ==> Authentication ==> Selected auth mode as OIDC.

we have filled all details as below screen shot(credential details are hidden state)

![harbor_OIDC_screen](https://user-images.githubusercontent.com/12935715/64937429-da7f6280-d877-11e9-84b3-2dfae9aa790f.PNG)

**Comment by reasonerjt** [comment]
I see, you setting is incorrect, you should remove the chunk `.well-known....` from your OIDC Endpoint.

**Comment by cmpmohan** [comment]
Its our OIDC dicovery URL  https://<domainname>/.well-known/openid-configuration. now its setup done without /.well-known/openid-configuration. receiving different error in harbor URL

![harbor_URL_errro](https://user-images.githubusercontent.com/12935715/64948265-51c2ef80-d894-11e9-8add-a278b22e9467.PNG)

**Comment by cmpmohan** [comment]
We have fixed 400 error message.  when we are clicking login via OIDC provider getting below error in core log.

=============================================

Failed to exchange token, error: oauth2: cannot fetch token: 400 Bad Request
Sep 17 23:57:30 172.24.0.1 core[16925]: Response: {"error_description":"code parameter is required for grant_type=authorization_code","error":"invalid_request"}
Sep 17 23:57:30 172.24.0.1 core[16925]: 2019-09-17T18:27:30Z [ERROR] [/common/api/base.go:69]: GET /c/oidc/callback?error_description=The+requested+scope+is+invalid%2C+unknown%2C+malformed%2C+or+exceeds+that+which+the+client+is+permitted+to+request.&state=a9ujbqfiaxd1u6ktb8njf7w4nl47q1dw&error=invalid_scope failed with error: {"code":400,"message":"oauth2: cannot fetch token: 400 Bad Request\nResponse: {\"error_description\":\"code parameter is required for grant_type=authorization_code\",\"error\":\"invalid_request\"}"}

**Comment by reasonerjt** [comment]
@cmpmohan again this is error in your configuration, the error is contained in the url of callback request:
```
Sep 17 23:57:30 172.24.0.1 core[16925]: 2019-09-17T18:27:30Z [ERROR] [/common/api/base.go:69]: GET /c/oidc/callback?error_description=The+requested+scope+is+invalid%2C+unknown%2C+malformed%2C+or+exceeds+that+which+the+client+is+permitted+to+request.&state=a9ujbqfiaxd1u6ktb8njf7w4nl47q1dw&error=invalid_scope failed with error: {"code":400,"message":"oauth2: cannot fetch token: 400 Bad Request\nResponse: {"error_description":"code parameter is required for grant_type=authorization_code","error":"invalid_request"}"}
```
The querystring for `error` reflects it:
```
error=invalid_scope
```

I suggest you talk to your admin or check the OIDC server's log to figure out what're the supported scopes and update the configuration in Harbor

**Comment by cmpmohan** [comment]
We have fixed supported scope section. now we are facing token expired issue. please find the below error. link https://auth0.com/docs/tokens/refresh-token/current saying we need to pass grant_type=refresh_token for token refresh.  how can i configure grant_type=refresh_token in harbor URL 

================================

Sep 18 16:11:50 172.27.0.1 core[16925]: 2019-09-18T10:41:50Z [ERROR] [/common/api/base.go:229]: oidc: token is expired (Token Expiry: 2019-09-17 06:49:01 +0000 UTC)
Sep 18 16:11:50 172.27.0.1 core[16925]: 2019-09-18T10:41:50Z [ERROR] [/common/api/base.go:69]: GET /c/oidc/callback?code=bUO8_vAZHAx7mqndAbHFXjvLNLo70RYm7DgAAADJ&state=84qbaak6pexlkbkxoo3d6dv4iwlub0as failed with error: {"code":500,"message":"Internal Server Error"}

**Comment by reasonerjt** [comment]
You don't need to do this, Harbor leverage `golang.org/x/oauth2` to implement this, your problem seems to me is that the initial token from auth0 is expired for some reason.

I have verifying against `auth0` on my radar, but unfortunately I don't have bandwidth this month, I plan to do it after the Oct holiday.

**Comment by cmpmohan** [comment]
do i need to implement as per docs https://github.com/coreos/go-oidc with my harbor to integrate OIDC?

**Comment by reasonerjt** [comment]
no, `go-oidc` is a library, actually Harbor uses it for token exchange.

**Comment by cmpmohan** [comment]
thanks for your response. Can you provide work around on this issue? or what is permanent fix?

**Comment by reasonerjt** [comment]
Based on the information you provided I really am not sure what's wrong.

In this log:
```
Sep 18 16:11:50 172.27.0.1 core[16925]: 2019-09-18T10:41:50Z [ERROR] [/common/api/base.go:229]: oidc: token is expired (Token Expiry: 2019-09-17 06:49:01 +0000 UTC)
Sep 18 16:11:50 172.27.0.1 core[16925]: 2019-09-18T10:41:50Z [ERROR] [/common/api/base.go:69]: GET /c/oidc/callback?code=bUO8_vAZHAx7mqndAbHFXjvLNLo70RYm7DgAAADJ&state=84qbaak6pexlkbkxoo3d6dv4iwlub0as failed with error: {"code":500,"message":"Internal Server Error"}
```

Looks like auth0 is returning a token that is expired already in token exchange when you log in, which is quite strange if the clock on your env is accurate.

**Comment by xaleeks** [comment]
@cmpmohan  I can help escalate this as we are trying to test against auth0.  We have already tested successfully against other providers.  Can you ping me some details on how you set this up on auth0?  please email xalex@vmware.com

**Comment by cmpmohan** [comment]
We have configured all OIDC credential in harbor URL. i have already shared log details above. so still if Auth0 is not tested or working with in OIDC setup. what option do i need to use other then auth0????

**Comment by cmpmohan** [comment]
I checked with our organization who is managing OIDC part. they are suggesting to follow below link to implement the harbor with OIDC.

 Golang implementation can be found there. So you could evaluate the use of this one:
1) https://github.com/coreos/go-oidc

If you have a web server in front of your backend, you could use the following module:

2) https://github.com/zmartzone/mod_auth_openidc

can you suggest if harbor is already included above setup or do i need to individual setup it as per above docs?

**Comment by reasonerjt** [comment]
@cmpmohan if you look at the code Harbor DOES use go-oidc for the interaction.

**Comment by reasonerjt** [comment]
@cmpmohan based on your comment, could you clarify whether it's true that you can login Harbor via Auth0 by OIDC, but see issue in refresh token?

I did some test today, seems Auth0 for some reason is not returning the id_token for refresh request, some user raised this:
https://community.auth0.com/t/missing-id-token-when-refreshing-tokens/13937

I'll try to reach out to Auth0 to see if there's any trick in the configuration.

**Comment by cmpmohan** [comment]
@reasonerjt ,  Harbor URL is able to redirect to OIDC page once clicked "OIDC login provider".  in login prompt when will provide login credential as username and password.. its throwing token expired error.

**Comment by reasonerjt** [comment]
How long is the lifetime of the token?  Is there any time mismatch?  You can gram the id_token in jwt format in Harbor's log, you can see the exp after decoding it.

**Comment by reasonerjt** [comment]
I believe this is fixed in #9435 , in which change Harbor no longer read the `id_token` field in response of token refresh.

I verified it on my auth0 env.

**Comment by cmpmohan** [comment]
where i can edit to disable the refresh_token ID in harbor? does this
token_refresh parameters available in harbor.yaml?

**Comment by cmpmohan** [comment]
> I believe this is fixed in #9435 , in which change Harbor no longer read the `id_token` field in response of token refresh.
> 
> I verified it on my auth0 env.

But where i need to edit harbor setup to change the refresh token field? could you please provide details explanation.

**Comment by adriannieto-attechnest** [comment]
Same here, could you help us providing a little bit more of detail?
